text/0000-union-initialization-and-drop.md

```diff
@@ -0,0 +1,367 @@
+(un)initializes its siblings.  For example:
+
+```rust
+// This code creates bad references and transmutes to `Vec` in incorrect ways.
+// This is just to demonstrate what the compiler would accept in terms of
+// tracking initialization.
+
+struct S(i32); // not `Copy`, no drop glue
+union U { f1: ManuallyDrop<Vec<i32>>, f2: (S, S), f3: i32 }
+
+let mut u: U;
+// Now `u` is not initialized: `&u`, `&u.f2` and `&u.f2.0` are all rejected.
+
+// We can write into uninitialized inner fields:
+u.f2.1 = S(42);
+{ let _x = &u.f2.1; } // This field is initialized now.
+// But this does not change the initialization state of the union itself,
+// or any other (inner) field.
+
+// We can initialize by assigning an entire field:
+u.f1 = ManuallyDrop::new(Vec::new());
+// Now *all (nested) fields* of `u` are initialized, including the siblings of `f1`:
+{ let _x = &u.f2; }
+{ let _x = &u.f2.0; }
+
+// Equivalently, we can assign the entire union:
+u = U { f2: (S(42), S(23) };
+// Now `u` is still initialized.
+
+// Copying does not change anything:
+let _x = u.f3;
+// Now `u` is still initialized.
+
+// We can move out of an initialized union:
+let v = u.f1;
+// Now `f1` *and its siblings* are no longer initialized (they got "moved out of"):
+// `let _x = u.f2;` would hence get rejected, as would `&u.f1` and `foo(u)`.
+u.f1 = v;
+// Now `u` and all of its fields are initialized again ("moving back in").
+
+// When we move out of an inner field, the other union fields become uninitialized
+// even if they are `Copy`.
+let s = u.f2.1;
+// Now `u.f1` and `u.f3` are no longer initialized.  But `u.f2.0` is:
+let s = u.f2.0;
+```
+
+If the union implements `Drop`, the same restrictions as for structs apply: It
+is not possible to initialize a field before initializing the entire variable,
+and it is not possible to move out of a field.  For example:
+
+```rust
+// This code creates bad references and transmutes to `Vec` in incorrect ways.
+// This is just to demonstrate what the compiler would accept in terms of
+// tracking initialization.
+
+struct S(i32); // not `Copy`, no drop glue
+
+union U { f1: ManuallyDrop<Vec<i32>>, f2: (S, S), f3: u32 }
+impl Drop for U {
+    fn drop(&mut self) {
+        println!("Goodbye!");
+    }
+}
+
+let mut u: U;
+// `u.f1 = ...;` gets rejected: Cannot initialize a union with `Drop` by assigning a field.
+u = U { f2: (S(42), S(1)) };
+// Now `u` is initialized.
+
+// `let v = u.f1;` gets rejected: Cannot move out of union that implements `Drop`.
+let v_ref = &mut u.f1; // creating a reference is allowed
+let _x = u.f3; // copying out is allowed
+```
+
+When a union implementing `Drop` goes out of scope, its destructor gets called if and only if the union is currently considered initialized:
+(Continuing the example from above.)
+
+```rust
+{
+    let u = U { f2: (S(0), S(1)) };
+    // drop gets called
+}
+{
+    let u = U { f1: ManuallyDrop::new(Vec::new()) };
+    foo(u.f2);
```

Comment:
This looks like moving out of a union that implements `Drop`-- how is this different from the `let v = u.f1;` case above?